Compile against typescript@next

alexeagle opened the pull request:
Uses an extra travis worker, but should complete very quickly.

Fixes #6368

btford commented:
@alexeagle what is the status of this?

alexeagle commented:
Ready for review - I'll need a ts2dart release including https://github.com/angular/ts2dart/pull/338 before this is green and mergable.

vsavkin commented on modules/angular2/src/common/pipes/async_pipe.ts:
```diff
@@ -33,6 +33,7 @@ class PromiseStrategy {
 
 var _promiseStrategy = new PromiseStrategy();
 var _observableStrategy = new ObservableStrategy();
+var __unused: Promise; // avoid unused import when Promise union types are erased
```
Not sure why this is required

vsavkin replied:
Aren't Observable and EventEmitter used?

alexeagle replied:
The problem is that Promise is only referenced in a union type, and ts2dart erases unions into `dynamic`. Then dart analyzer errors because of an unused import.
By making some non-erased reference to the type, we defeat the unused import check.

vsavkin commented:
Travis is red.

alexeagle commented:
Now that a ts2dart version bump is included in this commit, it should pass...

alexeagle committed:
ci(typescript): add typescript_next build

Install typescript@next before build.js and test.typings.
Restore the regular version before travis caches node_modules/.

Fixes #6368

alexeagle committed:
fix(typing): Remove re-export of the Promise built-in type.
Instead, ts2dart can add the 'dart:async' import whenever
Promise is used.

Fixes #6468

alexeagle committed:
chore(ts2dart): update ts2dart to 0.7.24

alexeagle commented:
attn caretaker:
To sync this to google3 you need to patch http://cl/114595197

mary-poppins commented:
Merging PR #6417 on behalf of @btford to branch [presubmit-btford-pr-6417](https://github.com/angular/angular/tree/presubmit-btford-pr-6417).

alexeagle commented:
The dart_ddc build failed on the presubmit run. I retried once, got the same result.
There are 5 new ddc errors, 11 rather than 6. To find the new ones, I diff'ed this result against a recent pass on master:
```
1d0
< severe: [AnalyzerMessage] The argument type '(bool) → dynamic' cannot be assigned to the parameter type '(Future<bool>) → dynamic' (package:angular2/src/router/router.dart, line 202, col 15)
6c5,11
< severe: [STATIC_TYPE_ERROR] Type check failed: (bool result) {if (!result) {return false;} return this._routerCanDeactivate(instruction).then((bool result) {if (result) {return this.commit(instruction, _skipLocationChange).then((_) {this._emitNavigationFinish(instruction.toRootUrl()); return true;});}});} ((bool) → dynamic) is not of type (Future<bool>) → dynamic (package:angular2/src/router/router.dart, line 202, col 15)
---
> severe: [INVALID_METHOD_OVERRIDE] Invalid override. The type of RawEntity.notifyPropertyChange ((Symbol, Object, Object) → dynamic) is not a subtype of ChangeNotifier.notifyPropertyChange (<T>(Symbol, T, T) → T). (src/observable_models/common.dart, line 141, col 3)
> severe: [STATIC_TYPE_ERROR] Type check failed: (List<List<String>> nestedStyles) {List<dynamic> result = plainStyles.map((plainStyle) => this._shimIfNeeded(plainStyle, encapsulate)).toList(); nestedStyles.forEach((styles) => result.add(styles)); return result;} ((List<List<String>>) → dynamic) is not of type (List<dynamic>) → dynamic (package:angular2/src/compiler/style_compiler.dart, line 86, col 46)
> severe: [STATIC_TYPE_ERROR] Type check failed: (Todo todo) => todo.completed ((Todo) → bool) is not of type (KeyModel) → bool (src/todo/index.dart, line 56, col 29)
> severe: [STATIC_TYPE_ERROR] Type check failed: (Todo todo) {todo.completed = isComplete;} ((Todo) → void) is not of type (KeyModel) → void (src/todo/index.dart, line 50, col 33)
> severe: [STATIC_TYPE_ERROR] Type check failed: isPresent ((dynamic) → bool) is not of type (Function) → bool (package:angular2/src/common/forms/validators.dart, line 104, col 46)
> severe: [STATIC_TYPE_ERROR] Type check failed: isPresent ((dynamic) → bool) is not of type (Function) → bool (package:angular2/src/common/forms/validators.dart, line 113, col 46)
> severe: [STATIC_TYPE_ERROR] Type check failed: mostSpecific ((List<Instruction>) → Instruction) is not of type (List<dynamic>) → Instruction (package:angular2/src/router/route_registry.dart, line 202, col 51)
```

I can't figure out why my change causes this, nor why I got a green result on travis for the same SHA.

alexeagle commented:
@tbosch says he observed the same thing on his PR - it seems that something changed in Dart that reports more DDC errors than before.
I increased the limit to 11 and merged this PR as a7e9bc9

jmesserly commented:
> @tbosch says he observed the same thing on his PR - it seems that something changed in Dart that reports more DDC errors than before.

Just a guess, since those are function types, we classify errors differently in a case like this:
```dart
typedef int ReturnsAnInt();
Object returnsAnyObject() { return 'hello'; }
main() {
  // error: this cast will never succeed.
  ReturnsAnInt f = returnsAnyObject;
  int x = f();
  print(x + 1);
}
```

Previously, that was a warning. It is now an error, because we know the exact type of "returnsAnyObject" (it is a named function), we know the downcast will never succeed. But this would just be a warning:

```dart
typedef int ReturnsAnInt();
Object returnsAnyObject() { return 'hello'; }
main() {
  var g = returnsAnyObject;
  // ...
  // warning: this cast probably won't succeed.
  ReturnsAnInt f = g;
  int x = f();
  print(x + 1);
}
```

If the parameter types are `dynamic` (TS: `any`), some relief is coming in https://codereview.chromium.org/1700523002/. That will fix a case like this:

```dart
takesAnything(x) {
  print(x);
}
typedef TakesAnInt(int x);
main() {
  // currently, this is an error.
  // after a fix, this will be allowed and will work at runtime.
  TakesAnInt t = takesAnything;
  t(42);
}
```

alexeagle commented:
Seems like a plausible explanation. Thanks!

On Wed, Feb 17, 2016 at 8:58 AM John Messerly <notifications@github.com>
wrote:

> @tbosch <https://github.com/tbosch> says he observed the same thing on
> his PR - it seems that something changed in Dart that reports more DDC
> errors than before.
>
> Just a guess, since those are function types, we classify errors
> differently in a case like this:
>
> typedef int ReturnsAnInt();
> Object returnsAnyObject() { return 'hello'; }
> main() {
>   // error: this cast will never succeed.
>   ReturnsAnInt f = returnsAnyObject;
>   int x = f();
>   print(x + 1);
> }
>
> Previously, that was a warning. It is now an error, because we know the
> exact type of "returnsAnyObject" (it is a named function), we know the
> downcast will never succeed. But this would just be a warning:
>
> typedef int ReturnsAnInt();
> Object returnsAnyObject() { return 'hello'; }
> main() {
>   var g = returnsAnyObject;
>   // ...
>   // warning: this cast probably won't succeed.
>   ReturnsAnInt f = g;
>   int x = f();
>   print(x + 1);
> }
>
> If the parameter types are dynamic (TS: any), some relief is coming in
> https://codereview.chromium.org/1700523002/. That will fix a case like
> this:
>
> takesAnything(x) {
>   print(x);
> }typedef TakesAnInt(int x);
> main() {
>   // currently, this is an error.
>   // after a fix, this will be allowed and will work at runtime.
>   TakesAnInt t = takesAnything;
>   t(42);
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/angular/angular/pull/6417#issuecomment-185296753>.
>
